src/imageproxy.rs

```diff
@@ -569,13 +612,67 @@ impl ImageProxy {
         let args: Vec<serde_json::Value> =
             vec![img.0.into(), digest.to_string().into(), size.into()];
         let (_bloblen, fd) = self.impl_request::<i64, _, _>("GetBlob", args).await?;
-        let (fd, pipeid) = fd.ok_or_else(|| Error::new_other("Missing fd from reply"))?;
-        let fd = tokio::fs::File::from_std(std::fs::File::from(fd));
+        let fd = fd.ok_or_else(|| Error::Other("Missing fd from reply".into()))?;
+        let FileDescriptors::FinishPipe { pipeid, datafd } = fd else {
+            return Err(Error::Other("got dualfds, expecting FinishPipe fd".into()));
+        };
+        let fd = tokio::fs::File::from_std(std::fs::File::from(datafd));
         let fd = tokio::io::BufReader::new(fd);
         let finish = Box::pin(self.finish_pipe(pipeid));
         Ok((fd, finish))
     }
 
+    async fn read_blob_error(fd: OwnedFd) -> std::result::Result<(), GetBlobError> {
+        let fd = tokio::fs::File::from_std(std::fs::File::from(fd));
+        let mut errfd = tokio::io::BufReader::new(fd);
+        let mut buf = Vec::new();
+        errfd
+            .read_to_end(&mut buf)
+            .await
+            .map_err(|e| GetBlobError::Other(e.to_string().into_boxed_str()))?;
+        if buf.is_empty() {
+            return Ok(());
+        }
+        #[derive(Deserialize)]
+        struct RemoteError {
+            code: String,
+            message: String,
+        }
+        let e: RemoteError = serde_json::from_slice(&buf)
+            .map_err(|e| GetBlobError::Other(e.to_string().into_boxed_str()))?;
+        match e.code.as_str() {
+            // Actually this is OK
+            "EPIPE" => Ok(()),
+            "retryable" => Err(GetBlobError::Retryable(e.message.into_boxed_str())),
+            _ => Err(GetBlobError::Other(e.message.into_boxed_str())),
+        }
+    }
+
+    /// Fetch a blob identified by e.g. `sha256:<digest>`; does not perform
+    /// any verification that the blob matches the digest. The size of the
+    /// blob and a pipe file descriptor are returned.
+    #[instrument]
+    pub async fn get_raw_blob(
+        &self,
+        img: &OpenedImage,
+        digest: &Digest,
+    ) -> Result<(
+        u64,
+        tokio::fs::File,
+        impl Future<Output = std::result::Result<(), GetBlobError>> + Unpin + '_,
+    )> {
+        tracing::debug!("fetching blob");
+        let args: Vec<serde_json::Value> = vec![img.0.into(), digest.to_string().into()];
+        let (bloblen, fd) = self.impl_request::<u64, _, _>("GetRawBlob", args).await?;
```

Comment:
So hear me out.  It seems like we call impl_request() on a bunch of different methods.  For each method, we're more or less expecting a specific response in terms of file descriptors, ya?

What if we made impl_request() also generic over the expected return type of the file descriptors in question?  Then we could TryFrom on the filtered cmsg data and the pipeid inside of the impl function, failing if that fails for our desired fd-result-type.

Then, by the time impl_request() returns, we already have (statically) our desired set of fds, pipes, whatever and we can remove all of the checks we're doing everywhere...

Reply:
So I spent about 15 minutes trying this and it got messy; maybe I was doing it wrong, but it exceeded my time budget I can work on a lot more invasive things here. So either we keep this as is, or someone else picks it up later, or we don't merge?

Reply:
I'll take a look